### PR TITLE
Add loading spinners and mermaid error handling

### PR DESCRIPTION
**Pull request by silverwind**
- Add loading spinners on editor and mermaid renderers
- Add error handling and inline error box for mermaid
- Fix mermaid rendering by using the `.init` API

image
image
image
image

### CODE REVIEW

**Comment by lafriks** [comment]
Could you rename `loadingcircle` to `is-loading`?

**Comment by silverwind** [comment]
I'm kind of wary of using too common classnames because they could clash with other frameworks like swagger-ui that uses `.loading`. `.is-loading` is probably fine thought.

**Comment by silverwind** [comment]
Classes switched.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Bit off topic... But related, what will happen if mermaid can not render image because of syntax error?

**Comment by silverwind** [comment]
I think it may throw an JS error, not sure how well that library behaves or if it has a fallback rendering. Looking at [GitLab code](https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/behaviors/markdown/render_mermaid.js), they seem to have a `.catch` handler for such an case.

**Comment by silverwind** [comment]
Added error handling for mermaid and a toast system that will pop up error boxes when mermaid fails to render. I think it will be useful for more JS features later on. Test with this:

````
```mermaid
    stateDiagram
    [*] --> Active

    state Active {
        [*] ---> NumLockOff
        NumLockOff --> NumLockOn : EvNumLockPressed
        NumLockOn --> NumLockOff : EvNumLockPressed
    }
```

**Comment by silverwind** [comment]
Maybe the toasts are actually overkill and a inline error would be enough. Need to think about it some more.

**Comment by silverwind** [comment]
Ready for review again. I went with a inline error block on mermaid errors:

**Comment by silverwind** [comment]
One more tweak: text is now left-aligned so the `----^` pointers work:

image

**Comment by silverwind** [comment]
Discovered an issue where mermaid.js seems to unavoidably spew an error rendering at the end of `<body>`. Pretty sure it's a bug in the library. Worked around by hiding it.

image
image

File: web_src/less/index.less [hunk 1]
```diff
@@ -1,5 +1,7 @@
 @import "~font-awesome/css/font-awesome.css";
 @import "./vendor/gitGraph.css";
+@import "./features/animations.less";
+@import "./markdown/mermaid.less";
```
Comment (zeripath) [inline web_src/less/index.less:4]: why not "./features/mermaid.less" ? There's nothing restricting mermaid to markdown...

Comment (silverwind) [reply web_src/less/index.less:4]: Right now it only appears in markdown and this file naming mirrors the one in `js`.

Comment (silverwind) [reply web_src/less/index.less:4]: Thought I guess if you prefer, we can move both to `features`, but standalone mermaid template files are not possible currently and it kind of breaks my concept of markdown content rendering (e.g. [this file](https://github.com/go-gitea/gitea/blob/master/web_src/js/markdown/content.js) which I specifically added so we can support more in-markdown content like math, videos and other stuff).

Comment (silverwind) [reply web_src/less/index.less:4]: Actually, I'm not going to change this. It's a markdown-only feature. Also in line with [gitlab's](https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/behaviors/markdown/render_mermaid.js) file naming scheme if that matters.

**Comment by silverwind** [comment]
It seems the way we initialize mermaid using `mermaidApi` is resulting in a weird/wrong renderings even thought docs say it's the recommended method. `mermaid.init` seems to produce better renderings, working on using that instead.

**Comment by silverwind** [comment]
Should be fixed. That API of mermaid is absolutely horrendous, especially it's error handling but this seems to produce much better rendering now for me.

**Comment by silverwind** [comment]
Squashed, update OP and removed a few outdated comments in this PR.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
I still think we'll end up with mermaid being used outside of markdown and will have to move stuff around but that's fine

**Comment by silverwind** [comment]
Yeah I guess we can move both less and js files around then but the current structure is good enough for markdown-only content like mermaid currently.

**Pull request merged**